File: doc-examples/hibernate-spring-tx-example-java/src/main/java/example/BookManager.java

package example;

import io.micronaut.context.event.ApplicationEventPublisher;
import io.micronaut.transaction.annotation.TransactionalEventListener;
import jakarta.inject.Singleton;
import org.springframework.transaction.support.TransactionSynchronizationManager;

import jakarta.transaction.Transactional;
import java.util.ArrayList;
import java.util.List;

@Singleton
public class BookManager {
    private final BookRepository bookRepository;
    private final ApplicationEventPublisher<NewBookEvent> eventPublisher;
    private final List<NewBookEvent> events = new ArrayList<>();

    public BookManager(BookRepository bookRepository, ApplicationEventPublisher<NewBookEvent> eventPublisher) {
        this.bookRepository = bookRepository;
        this.eventPublisher = eventPublisher;
    }

    @Transactional
    void saveBook(String title, int pages) {
        final Book book = new Book(title, pages);
        bookRepository.save(book);
        if (!TransactionSynchronizationManager.isActualTransactionActive()) {
            throw new IllegalStateException("No active Spring transaction!");
        }
        eventPublisher.publishEvent(new NewBookEvent(book));
    }

    @TransactionalEventListener
    void onNewBook(NewBookEvent event) {
        events.add(event);
    }

    public List<NewBookEvent> getEvents() {
        return events;
    }

    record NewBookEvent(Book book) {
    }
}

File: doc-examples/hibernate-example-python/src/main/python/example/BookManager.py
from dataclasses import dataclass

from jakarta.inject import Singleton
from jakarta.transaction import Transactional
from micronaut.context.event import ApplicationEventPublisher
from micronaut.transaction.annotation import TransactionalEventListener

from example.Book import Book
from example.BookRepository import BookRepository


@dataclass
class NewBookEvent:
    book: Book


@Singleton
class BookManager:

    def __init__(self, book_repository: BookRepository, event_publisher: ApplicationEventPublisher[NewBookEvent]):  # <1>
        self.book_repository = book_repository
        self.event_publisher = event_publisher

    @Transactional
    def save_book(self, title: str, pages: int) -> None:
        book = Book(title, pages)
        self.book_repository.save(book)
        self.event_publisher.publishEvent(NewBookEvent(book))  # <2>

    @TransactionalEventListener
    def on_new_book(self, event: NewBookEvent) -> None:
        print(f"book = {event.book}")  # <3>
